### PR TITLE
Allocation Change Flow

### PR DESCRIPTION
**Pull request by brisco17**
Resolves issue #294. This extends the Allocation system by allowing project PIs and Managers to request changes to project allocations and their attributes. 

Allocations and Allocation Attribute Types now have an "is_changeable" field that system admins can use to denote whether or not they want users to be able to request changes to those objects. If an Allocation is changeable, a user can request a date extension to that allocation. If an Allocation Attribute is changeable, users can request its value to be updated. 

To request changes, a PI should go to an Allocation's Detail Page and press the "Request Change" button in the "Allocation Information" table header which will take them to the "Change Request" page. This page takes the user's desired changes plus a justification for those changes to make an AllocationChangeRequest object (along with AllocationAttributeChangeRequest objects for every Allocation Attribute that has had a change requested, and a AllocationChangeStatusChoice object to denote the change's status).  The form also validates user requested Allocation Attribute values. 

Users can keep track of their Allocation change requests on a new "Allocation Change Requests" table on the Allocation's Detail page, which shows the date the change was requested, the change's status, any notes left by admins, and a link to a more detailed view of the request ("Allocation Change Detail" page). 

Admins can review change requests by selecting the 'Allocation Change Requests' link in the Admin tab of the navbar. The page is structured similarly to the 'Allocation Requests' page and Admins are given the option to either Approve/Deny changes on the spot, or enter their "Allocation Change Detail" pages where they can also leave notes for users. 

Users are notified by email when their changes are approved or denied, and once a change is activated the Allocation/Allocation Attributes are updated immediately.

### CODE REVIEW

**Comment by dsajdak** [comment]
On testing found:
1. The 'Request Change' button is available for all allocations whether they're marked 'is changeable' or not.  I think this should be displayed only if the allocation is changeable or if the default to make all allocations changeable is set.
2. A PI can request a change on an expired allocation which they shouldn't be able to do.  If the allocation is locked an error is displayed and the PI can not request the change, which is what we want for expired allocations too.
3. Marked an allocation 'is changeable' and logged in as the PI on that allocation.  Clicked 'Request Change,' selected '60 days' from the date extension drop down and entered a justification.  When I clicked 'submit' I saw this error:
error
NOTE: this happens whether I change the date or modify an attribute.  I also tried doing this as an admin and get the same error.

File: coldfront/core/allocation/templates/allocation/allocation_change.html [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  you must provide a justification.
+</p>
+
+  <div class="card mb-3">
+    <div class="card-header">
+        <h3><i class="fas fa-list" aria-hidden="true"></i> Allocation Information</h3>   
+    </div>
+
+    <div class="card-body">
+      <form action="{% url 'allocation-change' allocation.pk %}" method="post">
```
Comment (aebruno) [inline coldfront/core/allocation/templates/allocation/allocation_change.html:27]: The HTML is not valid after rendering this view. Looks like we have to move the `<form>` tag up a bit. Here's a patch that will fix the imbalanced tags:

```diff
--- a/coldfront/core/allocation/templates/allocation/allocation_change.html
+++ b/coldfront/core/allocation/templates/allocation/allocation_change.html
@@ -18,13 +18,13 @@ Request Allocation Change
   you must provide a justification.
 </p>
 
+<form action="{% url 'allocation-change' allocation.pk %}" method="post">
   <div class="card mb-3">
     <div class="card-header">
         <h3><i class="fas fa-list" aria-hidden="true"></i> Allocation Information</h3>   
     </div>
 
     <div class="card-body">
-      <form action="{% url 'allocation-change' allocation.pk %}" method="post">
         {% csrf_token %}
         <div class="table-responsive">
           <table class="table table-bordered table-sm">

```

**Review by aebruno — CHANGES_REQUESTED** [review]

**Comment by dsajdak** [comment]
All issues previously reported have been resolved.  Work flow looks good and all testing passed.  One final task is to make the allocation attribute changes editable by the admin.  They should be able to approve some but not all changes.  Add a garbage can icon to the allocation attribute table.  If the admin does not approve of the change request they can either modify the value or delete it entirely.  Add a modal for 'Are you sure you want to delete?'

For the date extension, add a '0 days' or 'No extension' option to the drop down for admins to select and then no change is made to the end date of the allocation.

**Review by dsajdak — APPROVED** [review]
All testing passed and this new allocation change process is complete.  Great contribution!

**Review by aebruno — APPROVED** [review]

**Pull request merged**